capreolus/extractor/bertpassage.py

```diff
@@ -232,27 +254,43 @@ def _prepare_doc_psgs(self, doc):
         """
         passages = []
         numpassages = self.config["numpassages"]
-        doc = self.tokenizer.tokenize(doc)
+        tokenizer = self.tokenizer.bert_tokenizer.backend_tokenizer
+        encoded_doc = tokenizer.encode(doc)
+        doc = encoded_doc.tokens[1:-1]
+
+        # For each wordpiece token, the beginning and end of the characters in the original doc
+        doc_offsets = encoded_doc.offsets[1:-1]
+        # doc = self.tokenizer.tokenize(doc)
+
+        # To get the word in the doc corresponding to a word in the passage:
+        # doc_offsets[passage_to_offsets[passage_id] + position of word in passage]
+        passage_to_begin_token = {}
 
         for i in range(0, len(doc), self.config["stride"]):
             if i >= len(doc):
                 assert len(passages) > 0, f"no passage can be built from empty document {doc}"
                 break
+
+            # Store the offset of the first character in the passage
+            passage_to_begin_token[len(passages)] = i
+
             passages.append(doc[i : i + self.config["passagelen"]])
 
         n_actual_passages = len(passages)
         # If we have a more passages than required, keep the first and last, and sample from the rest
         if n_actual_passages > numpassages:
             if numpassages > 1:
-                passages = [passages[0]] + list(self.rng.choice(passages[1:-1], numpassages - 2, replace=False)) + [passages[-1]]
+                # passages = [passages[0]] + list(self.rng.choice(passages[1:-1], numpassages - 2, replace=False)) + [passages[-1]]
```

Comment:
We might have to see how this affects the numbers - fold 1 ndcg_20 on Robust04 title queries with `electra-base-msmarco` was around 0.52, so this seems to be doing fine.